### DISCUSSION TITLE
Proposal: extend switch expression arms to allow multiple constant patterns

### DISCUSSION THREAD

**Discussion by IanKemp** [dotnet/csharplang#5087] [Language Ideas] [answered]
Currently a switch expression is limited to matching on a single constant pattern per case arm of the expression. This is incongruent with the switch statement, which allows for matching on multiple constant values in a single case via fall-through.

An example of a switch statement with fall-through:

```csharp
enum Foo
{
    Bar, Quux, Florb,
}

var foo = Foo.Florb;
string resultStatement;

switch (foo)
{
    case Foo.Bar:
        resultStatement = "bar";
        break;

    case Foo.Quux:
    case Foo.Florb:
        resultStatement = "baz";
        break;

    default:
        resultStatement = "unknown";
        break;
}
```

The equivalent switch expression to achieve the same functionality today - note the unnecessary duplication of the result value "baz" in the case of Quux and Florb:

```csharp
// (1)
var resultExpression = foo switch
{
    Foo.Bar => "bar",
    Foo.Quux => "baz",
    Foo.Florb => "baz",
    _ => "unknown",
}
```

It's possible to write the above as the following to eliminate the duplication via the use of case guards, but I have discounted this syntax as it feels like an abuse of the switch expression's intended simplicity:

```csharp
var resultExpression = foo switch
{
    Foo.Bar => "bar",
    _ when foo == Foo.Quux || foo == Foo.Florb => "baz",
    _ => "unknown",
}
```

I propose to extend switch expression syntax to allow the expression from (1) to be rewritten as something along the lines of:

```csharp
var resultExpression = foo switch
{
    Foo.Bar => "bar",
    Foo.Quux || Foo.Florb => "baz",
    _ => "unknown",
}
```

In other words, instead of each switch arm being limited to a single constant pattern, multiple constant patterns matching the type of the variable being switched on may be provided, separated by the `or` operator.

This would allow switch expressions to behave identically to switch statements with fall-through.

**Comment by YairHalberstadt** [comment, marked as answer]
You can currently do:

```csharp
var resultExpression = foo switch
{
    Foo.Bar => "bar",
    Foo.Quux or Foo.Florb => "baz",
    _ => "unknown",
}
```

**Reply by IanKemp** [reply]
Oh gosh darn it. Patterns using English words for comparison operators instead of the comparison operators that everyone knows or loves, claims yet another victim. For anyone else's reference: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/patterns#logical-patterns (C# 9.0+ only).

**Reply by IanKemp** [reply]
@YairHalberstadt Do you think it would be helpful to modify https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/switch-expression to call out the fall-through equivalent using pattern combinators? Since this is IMO quite a common case where devs might look at the switch expression, go "it doesn't have fall-through so we can't use it", and continue using the unnecessarily verbose switch statement.

**Reply by YairHalberstadt** [reply]
I guess so, but that's not really my area - I'm just a community member here.

**Reply by 333fred** [reply]
It's a good idea. @BillWagner?
